tools/codegen/gen.py

```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
+
+                out_impl_name = f"at::native::{g.out.dispatch[dispatch]}"
+
+                # TODO: work a little harder to generate fresh names here
+                # TODO: less praying that I picked the right argument names
+
+                if k is SchemaKind.functional:
+                    out_expr = "result"
+                    prologue = "auto result = tensor_from_meta(meta_result);"
+                elif k is SchemaKind.inplace:
+                    out_expr = "self"
+                    prologue = "// TODO: consistency check assert"
+                elif k is SchemaKind.out:
+                    # TODO: generalize this for multi-out
+                    assert len(f.func.out_arguments) == 1, "multi-out structured not supported yet"
+                    # TODO: naughtily leaning on some codegen invariants....
+                    out_expr = f.func.out_arguments[0].name
+                    prologue = f"""
+// TODO: add a consistency check for meta_result
+{out_expr}.resize_(meta_result.sizes);
+"""
+
+                device_guard = ""
+
+                if dispatch is None or 'CUDA' in dispatch or 'Vulkan' == dispatch:
+                    # TODO: stop copypaste
+                    self_args = (a for a in f.func.arguments if a.name == "self")
+                    candidate_args = itertools.chain(self_args, f.func.out_arguments, f.func.arguments)
+                    device_of = next((f'{a.name}' for a in candidate_args if a.type.is_tensor_like()), None)
+
+                    device_guard = ''
+                    if f.device_guard and device_of is not None:
+                        # TODO: OptionalCUDAGuard
+                        device_guard = f"""\
+    const OptionalDeviceGuard device_guard(device_of({device_of}));
+"""
+
+                return f"""\
+{sig.defn()} {{
+    {device_guard}
+    auto meta_result = meta::{meta_name}({functional_exprs});
+    {prologue}
+    {out_impl_name}({out_expr}, {functional_exprs});
+    return {out_expr};
+}}
+"""
+
+            elif target is Target.REGISTRATION:
+                if dispatch is None:
+                    return f'm.def({cpp_string(str(f.func))});'
+                else:
+                    type_name = qualify_type_name(sig.name())
+                    if local.use_c10_dispatcher() is UseC10Dispatcher.full:
+                        return f'm.impl("{f.func.name}", TORCH_FN({type_name}));'
+                    else:
+                        # SIGHHHHH
+                        return f'm.impl("{f.func.name}", torch::CppFunction::makeUnboxedOnly({type_name}));'
+            else:
+                assert_never(target)
+
+        rs = []
+        rs.append(subfunc(g.out))
+        if g.functional is not None:
+            rs.append(subfunc(g.functional))
+        if g.inplace is not None:
+            rs.append(subfunc(g.inplace))
+
+        return rs
+
+    def group_func(g: NativeFunctionGroup) -> List[str]:
+        if g.structured():
+            return structured_func(g)
```

Comment:
IIUC the dispatch key checks [here](https://github.com/pytorch/pytorch/pull/45277/files#diff-6382c847dac2d011f822c492e61f9fa0eceb47c2d6fc5fb4f0c42af726c1a3aaR348-R352) might kick us back out with `[]`, is that ok downstream?

Reply:
Err, yes? The whole point of returning [] sometimes is that sometimes you don't have a registration for a given dispatch key

Reply:
Ah got it - I was reading "unsupported" in the called function as meaning unsupported by structured kernels specifically - forgot that this gets called for literally all dispatch keys. All caught up now 😁